### PR TITLE
Introduce trafficserver module

### PR DESCRIPTION
**Pull request by paladox**
Install Apache Traffic Server a replacement for varnish.

The module is from [1] but modified for our needs.

[1] https://github.com/wikimedia/puppet/tree/production/modules/trafficserver

### CODE REVIEW

File: hieradata/hosts/test4.yaml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      params:
+          - '@plugin=/usr/lib/trafficserver/modules/tslua.so'
+          - '@pparam=/etc/trafficserver/lua/normalize-path.lua'
+          # decode    :  /  @  !  $  (  )  *  ,  ;
+          - '@pparam="3A 2F 40 21 24 28 29 2A 2C 3B"'
+          # encode    [  ]  &  '  +  =
+          - '@pparam="5B 5D 26 27 2B 3D"'
+          - '@plugin=/usr/lib/trafficserver/modules/tslua.so'
+          - '@pparam=/etc/trafficserver/lua/rb-mw-mangling.lua'
+          - '@plugin=/usr/lib/trafficserver/modules/tslua.so'
```
Comment (Southparkfan) [inline hieradata/hosts/test4.yaml:22]: A lot of extra config. Is each item necessary?

File: hieradata/hosts/test4.yaml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    enable_tlsv1_3: 1
+  max_record_size: 16383
+  session_cache: 2
+  session_cache_timeout: 86400
+  session_cache_auto_clear: 1
+  session_cache_size: 512000
+  session_cache_buckets: 32768
+  session_ticket_enable: 0
+  load_elevated: 1
+  do_ocsp: 0
```
Comment (Southparkfan) [inline hieradata/hosts/test4.yaml:61]: We currently do OCSP stapling in NGINX.

File: hieradata/hosts/test4.yaml [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    # ClientTTFB is the time spent between the very beginning of this session
+    # and when the response header write to the client starts.
+    format: 'Date:%<cqtd> Time:%<cqtt> ConnAttempts:%<sca> ConnReuse:%<sstc> TTFetchHeaders:%<{TS_MILESTONE_SERVER_FIRST_READ-TS_MILESTONE_SERVER_FIRST_CONNECT}msdms> ClientTTFB:%<{TS_MILESTONE_UA_BEGIN_WRITE-TS_MILESTONE_SM_START}msdms> CacheReadTime:%<{TS_MILESTONE_CACHE_OPEN_READ_END-TS_MILESTONE_CACHE_OPEN_READ_BEGIN}msdms> CacheWriteTime:%<{TS_MILESTONE_CACHE_OPEN_WRITE_END-TS_MILESTONE_CACHE_OPEN_WRITE_BEGIN}msdms> TotalSMTime:%<{TS_MILESTONE_SM_FINISH-TS_MILESTONE_SM_START}msdms> OriginServer:%<shn> OriginServerTime:%<stms> CacheResultCode:%<crc> CacheWriteResult:%<cwr> ReqMethod:%<cqhm> RespStatus:%<pssc> OriginStatus:%<sssc> ReqURL:%<cquuc> ReqHeader:User-Agent:%<{User-agent}cqh> ReqHeader:Host:%<{Host}cqh> ReqHeader:X-Client-IP:%<{X-Client-IP}cqh> ReqHeader:Cookie:%<{Cookie}cqh> BerespHeader:Set-Cookie:%<{Set-Cookie}ssh[0:16]> BerespHeader:Cache-Control:%<{Cache-Control}ssh> BerespHeader:Connection:%<{Connection}ssh> RespHeader:X-Cache-Int:%<{X-Cache-Int}psh> RespHeader:Backend-Timing:%<{Backend-Timing}psh>'
```
Comment (Southparkfan) [inline hieradata/hosts/test4.yaml:84]: Note: at the moment X-Real-IP is used on destination servers for determining the correct source IP.

File: modules/role/files/trafficserver/default.lua [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    -- We should also set proxy.config.cache.enable_read_while_writer to 0 but
+    -- there seems to be no TS_LUA_CONFIG_ option for it.
+end
+
+function no_cache_lookup()
+    ts.http.config_int_set(TS_LUA_CONFIG_HTTP_CACHE_HTTP, 0)
+end
+
+function do_global_read_request()
+    if ts.client_request.header['Host'] == 'healthcheck.miraheze.org' and ts.client_request.get_uri() == '/ats-be' then
```
Comment (Southparkfan) [inline modules/role/files/trafficserver/default.lua:65]: Do we even use this health check?

File: modules/trafficserver/files/rsyslog.conf [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Send certain ATS syslog output to logstash
```
Comment (Southparkfan) [inline modules/trafficserver/files/rsyslog.conf:1]: We don't use this?

File: modules/trafficserver/templates/logrotate.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<%= @paths['logdir'] %>/*.log {
```
Comment (Southparkfan) [inline modules/trafficserver/templates/logrotate.erb:1]: General note: ATS should log to syslog directly. In syslog-ng we can decide whether to forward to graylog, local log files, both, etc.

**Review by Southparkfan — CHANGES_REQUESTED** [review]

File: hieradata/hosts/test4.yaml (hunk 1 above)
Comment (paladox) [reply hieradata/hosts/test4.yaml]: Removed, thanks, done.

File: hieradata/hosts/test4.yaml (hunk 2 above)
Comment (paladox) [reply hieradata/hosts/test4.yaml]: Removed, thanks, done.

File: modules/trafficserver/files/rsyslog.conf (hunk 5 above)
Comment (paladox) [reply modules/trafficserver/files/rsyslog.conf]: Removed.

File: modules/trafficserver/templates/logrotate.erb (hunk 6 above)
Comment (paladox) [reply modules/trafficserver/templates/logrotate.erb:1]: ATS only supports sending system and emergencies to syslog.

See https://docs.trafficserver.apache.org/en/9.0.x/admin-guide/logging/destinations.en.html#syslog

**Comment by paladox** [comment]
Per conversion with SPF we've decided to merge.

**Pull request merged**